ast/node.go

```diff
@@ -377,6 +370,54 @@ func (c *CallExpr) Equal(other Node) bool {
 	return c.Callee.Equal(o.Callee)
 }
 
+// NewFunDecl creates a new function declaration node.
+func NewFunDecl(name Ident, args []Ident, body *Program) *FunDecl {
+	return &FunDecl{
+		Name: name,
+		Args: args,
+		Body: body,
+	}
+}
+
+func (a *FunDecl) Type() NodeType {
+	return NodeFunDecl
+}
+
+func (a *FunDecl) String() string {
+	var args []string
```

Comment:
we already had this conversation. For me, Go initializes slice and maps in the same way (empty value is nil for both) but usage of both are completely different =)

```go
var a []string
a[10] // panic: index out of bounds
a[20] = "test" // panic: index out of bounds

in slices, read and write panics.
```
append is another kind of funtion (that only makes sense for slices) that initializes the slice if it's not initialized.

var m map[string]interface{}
m["test"] // returns nil
m["test"] = 1 // panic: assign to nil map

One way to resemble 'append' in maps:
https://play.golang.org/p/yOwVLcbl0tb

I always treat map and slice different. And depending on the code, I initialize an slice when creating the variable (as you proposed) and sometimes not.

Using `var args []string` nothing is allocated in case len(a.Args) == 0. I'm not sure Go is smart enough to optimize the explicit allocation to NO OP in this same case.

Reply:
I remember our conversation, was just reviewing stuff in my mind, sorry to clutter the review with non-sensical stuff.